Doubly-linked list implementation

NikiNatov opened the pull request:
Doubly-linked list implementation using sentinel nodes.

NikiNatov committed:
Doubly-linked list implementation

mchechev commented on Examples/c++/Linked List/DLList.h:
```diff
@@ -0,0 +1,401 @@
+#include <algorithm>
+#include <utility>
+
+#ifdef DEBUG
```
It's a great style to use assert, but I believe it's not standart for the course, so please add few comments which are going to help your collegues to undestand why you are doing this. This also could be done and at several other parts of the code below.

NikiNatov replied:
I agree. Comments will be added explaining what the assert macro does.

mchechev commented on Examples/c++/Linked List/DLList.h:
```diff
@@ -0,0 +1,401 @@
+		using RefType = ValueType&;
+	public:
+		DLLIterator()
+			: m_Ptr(nullptr), m_SentinelNode(nullptr)
```
why naming with a preffix and underscore? everything above is named cammelCase

NikiNatov replied:
I use m_ to mark my member/private variables only. The names of the variables will be changed to be more sensible and discribing.

mchechev commented on Examples/c++/Linked List/DLList.h:
```diff
@@ -0,0 +1,401 @@
+		using RefType = ValueType&;
+	public:
+		DLLIterator()
+			: m_Ptr(nullptr), m_SentinelNode(nullptr)
```
also Ptr and SentinelNode are probably not the best names...

mchechev commented on Examples/c++/Linked List/DLList.h:
```diff
@@ -0,0 +1,401 @@
+	class DLList
+	{
+	public:
+		using ValueType = T;
```
are this usings really needed, I dont see them actually improving the readibility.

NikiNatov replied:
Some of the usings are indeed not used and will be removed from the code.

mchechev commented on Examples/c++/Linked List/DLList.h:
```diff
@@ -0,0 +1,401 @@
+			DeleteNode(where);
+		}
+
+		void MoveToBack(Iterator from, Iterator to)
```
It's not a standart function. Its not very clear why it's included at the list and also the name is not biving the reader perfect understanding what the function is going to do. The minimum for this function is to add a comment to it.

NikiNatov replied:
I left it there as a utility function but it can be safely removed from the list.

mchechev commented on Examples/c++/Linked List/DLList.h:
```diff
@@ -0,0 +1,401 @@
+			m_SentinelNode->Next->Previous = m_SentinelNode;
+		}
+
+		void Reverse(Iterator from, Iterator to)
```
Why having several functions which are doing the same without reusing each other. It's making the code difficult to maintain as if a bug is found at one of the function the other should also be checked

NikiNatov replied:
The Reverse() function override with no parameters is not necessarily needed so it will be removed.

mchechev commented on Examples/c++/Linked List/DLList.h:
```diff
@@ -0,0 +1,401 @@
+			beginNode->Next = endNodeNext;
+		}
+
+		void Clear()
```
Do we need this function public?

NikiNatov replied:
This works the same as the STL clear functions. I left it public so if the user wants to clear the whole list they have that option.

mchechev commented on Examples/c++/Linked List/DLList.h:
```diff
@@ -0,0 +1,401 @@
+			InsertBefore(m_SentinelNode, data);
+		}
+
+		void PushBack(T&& data)
```
Why having this 2 functions?

NikiNatov replied:
The second function takes rvalue reference as parameter. This allowes us to use move semantics which significantly improves the performance when we use heavy object types and we do not need to copy the object when passing it to the function.

mchechev commented on Examples/c++/Linked List/DLList.h:
```diff
@@ -0,0 +1,401 @@
+			m_Size++;
+		}
+
+		template<typename... Args>
```
For me these duplicates of the functions should not be done and some more comments should be added. Why actually this two functions even have a different behaviour. Could you explain a bit more?

NikiNatov replied:
This is also important for object types. Similar to the STL function emplace and emplace_back this function constructs the object in-place given its parameters which is also faster than copying it.

NikiNatov committed:
Updated Doubly-linked list

NikiNatov commented:
* Updates made *
-Assert macro explained
-Removed unnecessary functions
-Added comments on each function
-Renamed variable names for more clear understanding of the code
-Removed unnecessary usings
-Explained why certain function overrides exist and what they do
